### ISSUE TITLE
Too much inlining

### ISSUE THREAD

**Issue by mikesperber** [conal/concat#50]
When we uncomment this line:

https://github.com/conal/concat/blob/master/examples/test/Examples.hs#L421

i.e. this one:

`runCirc "affRelu"         $ toCcc $                affRelu @(Vector 2) @(Vector 3) @R`

The plugin fails to transform the `toCCC` away. Looking at the plugin output, it appears the problem is that the `runCirc` gets inlined to an insane amount of IO-related code: Without the call to `runCirc`, it works fine.

Unfortunately, turning the `INLINE` annotation for `runCirc` into `NOINLINE` doesn't change anything - the inlining still happens.

**Comment by conal** [comment]
I think I'm seeing the same behavior, but the problematic inlined code is from the vector library. I use the *vector-sized* library, which is a size-typed layer on top of the *vector* library. The former inlines aggressively, and the latter generates a ton of code that my plugin can't handle. This one issue in particular has been quite a struggle.

Ultimately, I'm aiming at running insanely fast on top of GPUs, FPGAs, etc, leaving the *vector* library behind, but for now it's a stepping stone. My general strategy is to capture code early enough that the problematic *vector* code never arises, but in that way I'm going against the flow of GHC, which generally leads to difficulty.

**Comment by conal** [comment]
I think I'm seeing the same behavior, but the problematic inlined code is from the vector library (monadic but not IO). I use the *vector-sized* library, which is a size-typed layer on top of the *vector* library. The former inlines aggressively, and the latter generates a ton of code that my plugin can't handle. This one issue in particular has been quite a struggle.

Ultimately, I'm aiming at running insanely fast on top of GPUs, FPGAs, etc, leaving the *vector* library behind, but for now it's a stepping stone. My general strategy is to capture code early enough that the problematic *vector* code never arises, but in that way I'm going against the flow of GHC, which generally leads to difficulty.

In my tracing output for this example, I see `affRelu` immediately inlining to a big chunko f code that includes a call to `foldrM` over an *unsized* vector and from which the plugin will not be able to recover. 
The `INLINE` pragmas I use so much are suppose to ensure that GHC inlines definitions as they are, not optimized forms of those definitions. This behavior is crucial for preventing problematic code such as in the vector library getting inlined and optimized before my plugin can intervene. In this example, problematic vector code is getting in there anyway. I just tried some simpler examples (from picking apart the definition of `affRelu`):

``` haskell
  , runCirc "foo" $ toCcc $ D.linear @(Vector 2) @(Vector 3) @R

  , runCirc "foo" $ toCcc $ (D.<.>) @(Vector 2) @R

  , runCirc "foo" $ toCcc $ sum @(Vector 2) @R
```

Like `affRelu`, these examples expose unsized vector code and then fail. I *thought* these examples all worked fine when I tried them last, but I guess I changed something and didn't test the `Deep` examples.

Ah. It seems that I switched from `sum` (assuming `Num`) to `sumA` (assuming `Additive`) at some point but didn't fix `(<.>)` in `Deep` according. Just fixed.
